Text widget annotations: implement comb support

timvandermeij opened the pull request:
This is another part of #7613.

timvandermeij committed:
Text widget annotations: implement comb support

Snuffleupagus commented on src/core/annotation.js:
```diff
@@ -715,6 +715,8 @@ var TextWidgetAnnotation = (function TextWidgetAnnotationClosure() {
     this.data.readOnly = this.hasFieldFlag(AnnotationFieldFlag.READONLY);
     this.data.multiLine = this.hasFieldFlag(AnnotationFieldFlag.MULTILINE);
+    this.data.comb = this.hasFieldFlag(AnnotationFieldFlag.COMB) &&
+                     this.data.maxLen !== null && !this.data.multiLine;
```
I think that you're missing a number of field checks here, please see the specification text:
> (PDF 1.5) May be set only if the MaxLen entry is present in the text field dictionary (see Table 229) and ***if the Multiline, Password, and FileSelect flags are clear***. If set, the field shall be automatically divided into as many equally spaced positions, or combs, as the value of MaxLen, and the text is laid out into those combs.

timvandermeij replied:
Good point. I thought we didn't need it because we do not support the `Password` and `FileSelect` flags anyway, but it makes sense to add these checks anyway to prevent any issues. I'll update the patch.

Snuffleupagus commented on web/annotation_layer_builder.css:
```diff
@@ -77,6 +77,22 @@
   border: 1px solid transparent;
 }
 
+.annotationLayer .textWidgetAnnotation input.comb {
```
Don't you want to add these rules to the corresponding file in `/test` as well, to enable testing of `comb` fields once we have better tests?

timvandermeij replied:
Yes, I'll do that.

timvandermeij commented:
I have addressed the comments and managed to adjust the reference test for text widget annotations to include a field that uses combs.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio test

timvandermeij commented:
I pushed a new commit to fix an issue in the a unit test. No other changes have been made.

/botio unittest

Snuffleupagus commented on test/unit/annotation_layer_spec.js:
```diff
@@ -499,13 +500,18 @@ describe('Annotation layer', function() {
         function() {
+      var flags = 0;
+      flags |= 1 << (AnnotationFieldFlag.READONLY - 1);
+      flags |= 1 << (AnnotationFieldFlag.MULTILINE - 1);
```
As far as I'm concerted, the current way that the `AnnotationFieldFlag` are defined makes the unit-tests unnecessarily convoluted and difficult to understand.

What would be *a lot* clearer is if you changed `AnnotationFieldFlag` to the same form as `AnnotationFlag`, since that means you could change the `hasFieldFlags` method in `src/core/annotation.js` to just:
```javascript
return !!(this.data.fieldFlags & flag);`
```
and specifying the flags in the unit-tests would be so much more straightforward:
```javascript
var flags = AnnotationFieldFlag.READONLY + AnnotationFieldFlag.MULTILINE;
```

Snuffleupagus replied:
As far as I'm concerted, the current way that the `AnnotationFieldFlag` are defined makes the unit-tests unnecessarily convoluted and difficult to understand.

What would be *a lot* clearer is if you changed `AnnotationFieldFlag` to the same form as `AnnotationFlag`, since that means you could change the `hasFieldFlags` method in `src/core/annotation.js` to just:
```javascript
return !!(this.data.fieldFlags & flag);`
```
and specifying the flags in the unit-tests would be so much more straightforward:
```javascript
var flags = AnnotationFieldFlag.READONLY + AnnotationFieldFlag.MULTILINE;
```

timvandermeij replied:
Really good point. I'll definitely look into that.

timvandermeij replied:
I pushed a new commit that fixes this.

Snuffleupagus commented on web/annotation_layer_builder.css:
```diff
@@ -77,6 +77,22 @@
+   * area, causing horizontal scrolling. We avoid this by extending the width
+   * when the element has focus and revert this when it loses focus.
+   */
+  width: 115%;
```
It's unfortunate that this sort of hack is necessary, but I'm also not able to come up with a better solution :-(

timvandermeij replied:
I agree, but after searching for a couple of hours there really isn't another way with pure CSS only. The real problem is that letter spacing is always applied on the right side of a character, whereas we would want to tune that. Since that's not possible, we'll need this workaround. There are JS-based solutions, but they are even uglier and require much more code, so this was the cleanest way to go.

Snuffleupagus replied:
I just want to make it absolutely clear that I didn't mean to complain about your solution here, but rather about the (missing) features of the web platform that made the workaround necessary!

Snuffleupagus commented on test/unit/annotation_layer_spec.js:
```diff
@@ -518,6 +524,74 @@ describe('Annotation layer', function() {
+      expect(textWidgetAnnotation.data.comb).toEqual(true);
+    });
+
+    it('should only accept comb fields when the flags are valid', function() {
```
It might be possible to simplify this particular test a bit, so I'd like to see an updated version based on the `AnnotationFieldFlag` comment above.

timvandermeij replied:
I'll try that. I'm not sure how much simpler it will become though as we still need the loops to remove the flags one-by-one, but we'll see.

timvandermeij replied:
I pushed a new commit that fixes this.

Snuffleupagus reviewed (changes requested):
Looks mostly fine, but I think that the format of `AnnotationFieldFlag` is making things more difficult to follow. Please see inline comments.

timvandermeij committed:
Widget annotations: simplify field flag handling

Directly use the hexadecimal representation, just like the
`AnnotationFlags`, to avoid calculations and to improve readability.
This allows us to simplify the unit tests for text widget annotations as
well.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio test

Snuffleupagus commented:
Much better, thank you! With the changes to `AnnotationFieldFlags`, the code and especially the unit-tests are now much simpler to deal with.

/botio makeref
